File: Assets/Module1_DataModels_and_IO/Scripts/ScenarioGeneration/DataModels/ScenarioEnums.cs
// =============================================================================
// ScenarioEnums.cs
// Module 1 – Dynamic Scenario Generation
// Team Sentinels | University of Moratuwa | 2026
// Unity 2022.3 LTS | Meta Quest 3
//
// Defines all enumeration types used across ScenarioConfig and Scenario data
// models. String-valued enums are serialised via Newtonsoft.Json's
// StringEnumConverter to match the JSON schema's snake_case string values.
// =============================================================================

using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using System.Runtime.Serialization;

namespace TeamSentinels.ScenarioGeneration.DataModels
{
    // =========================================================================
    // Mission Structure Enums
    // =========================================================================

    /// <summary>
    /// Mission template identifier. Currently only hostage_rescue is supported;
    /// reserved for future expansion to other tactical scenario types.
    /// </summary>
    [JsonConverter(typeof(StringEnumConverter))]
    public enum MissionType
    {
        [EnumMember(Value = "hostage_rescue")]
        HostageRescue
    }

    /// <summary>
    /// Room dimension category mapped to Unity world units.
    /// small = 4×4 m, medium = 6×6 m, large = 8×8 m.
    /// </summary>
    [JsonConverter(typeof(StringEnumConverter))]
    public enum RoomSizeCategory
    {
        [EnumMember(Value = "small")]
        Small,

        [EnumMember(Value = "medium")]
        Medium,

        [EnumMember(Value = "large")]
        Large
    }

    /// <summary>
    /// Room graph topology constraint.
    /// linear = sequential chain; branching = tree with decision points;
    /// hub_and_spoke = central room with radiating corridors;
    /// loop = cyclical paths allowing flanking.
    /// </summary>
    [JsonConverter(typeof(StringEnumConverter))]
    public enum LayoutType
    {
        [EnumMember(Value = "linear")]
        Linear,

        [EnumMember(Value = "branching")]
        Branching,

        [EnumMember(Value = "hub_and_spoke")]
        HubAndSpoke,

        [EnumMember(Value = "loop")]
        Loop
    }

    /// <summary>
    /// Number of building entry points.
    /// single = one breach point; multiple = 2–3 entry points.
    /// </summary>
    [JsonConverter(typeof(StringEnumConverter))]
    public enum EntryType
    {
        [EnumMember(Value = "single")]
        Single,

        [EnumMember(Value = "multiple")]
        Multiple
    }

    // =========================================================================
    // Entity Configuration Enums
    // =========================================================================

    /// <summary>
    /// Spatial distribution pattern for entity placement.
    /// clustered = entities in adjacent rooms (compounding tactical effects [5]);
    /// dispersed = spread across layout; front_loaded = near entry;
    /// deep = far from entry in high-depth rooms.
    /// </summary>
    [JsonConverter(typeof(StringEnumConverter))]
    public enum PlacementStrategy
    {
        [EnumMember(Value = "clustered")]
        Clustered,

        [EnumMember(Value = "dispersed")]
        Dispersed,

        [EnumMember(Value = "front_loaded")]
        FrontLoaded,

        [EnumMember(Value = "deep")]
        Deep
    }

    /// <summary>
    /// Hostage-terrorist proximity constraint level using distance thresholds [16].
    /// high = small max distance (closely guarded);
    /// medium = moderate spacing;
    /// low = large min distance (separated from threats).
    /// </summary>
    [JsonConverter(typeof(StringEnumConverter))]
    public enum HostageRiskLevel
    {
        [EnumMember(Value = "low")]
        Low,

        [EnumMember(Value = "medium")]
        Medium,

        [EnumMember(Value = "high")]
        High
    }

    /// <summary>
    /// Structural variation between generated scenarios [8][13].
    /// low = minimal variation from base topology;
    /// medium = moderate randomisation;
    /// high = maximum variation in room positions, sizes, and connectivity.
    /// </summary>
    [JsonConverter(typeof(StringEnumConverter))]
    public enum RandomnessLevel
    {
        [EnumMember(Value = "low")]
        Low,

        [EnumMember(Value = "medium")]
        Medium,

        [EnumMember(Value = "high")]
        High
    }

    // =========================================================================
    // Scenario Output Enums
    // =========================================================================

    /// <summary>
    /// Room type derived from generation context.
    /// entry = building entry room; corridor = connecting passage;
    /// standard = general-purpose room; hostage_room = room containing the hostage.
    /// </summary>
    [JsonConverter(typeof(StringEnumConverter))]
    public enum RoomType
    {
        [EnumMember(Value = "entry")]
        Entry,

        [EnumMember(Value = "corridor")]
        Corridor,

        [EnumMember(Value = "standard")]
        Standard,

        [EnumMember(Value = "hostage_room")]
        HostageRoom
    }

    /// <summary>
    /// Wall on which a door is placed.
    /// </summary>
    [JsonConverter(typeof(StringEnumConverter))]
    public enum WallSide
    {
        [EnumMember(Value = "north")]
        North,

        [EnumMember(Value = "south")]
        South,

        [EnumMember(Value = "east")]
        East,

        [EnumMember(Value = "west")]
        West
    }

    /// <summary>
    /// Entity actor type used across all modules for stable identification.
    /// </summary>
    [JsonConverter(typeof(StringEnumConverter))]
    public enum EntityType
    {
        [EnumMember(Value = "trainee")]
        Trainee,

        [EnumMember(Value = "hostage")]
        Hostage,

        [EnumMember(Value = "terrorist")]
        Terrorist
    }

    /// <summary>
    /// NPC role type assigned based on spatial structure and difficulty [18].
    /// patrol = traverses a circuit of rooms;
    /// stationary_guard = holds a fixed position;
    /// roaming_guard = moves within a bounded set of adjacent rooms;
    /// hostage_guardian = guards the hostage directly.
    /// </summary>
    [JsonConverter(typeof(StringEnumConverter))]
    public enum NpcRole
    {
        [EnumMember(Value = "patrol")]
        Patrol,

        [EnumMember(Value = "stationary_guard")]
        StationaryGuard,

        [EnumMember(Value = "roaming_guard")]
        RoamingGuard,

        [EnumMember(Value = "hostage_guardian")]
        HostageGuardian
    }

    /// <summary>
    /// Navigation context type discriminator, matching the role for polymorphic
    /// deserialisation of navigation context entries.
    /// </summary>
    [JsonConverter(typeof(StringEnumConverter))]
    public enum NavigationContextType
    {
        [EnumMember(Value = "patrol")]
        Patrol,

        [EnumMember(Value = "stationary_guard")]
        StationaryGuard,

        [EnumMember(Value = "roaming_guard")]
        RoamingGuard,

        [EnumMember(Value = "hostage_guardian")]
        HostageGuardian
    }
}

File: sentinels-aar/lib/scenarioEnums.js
// Mirrors Module 1's ScenarioEnums.cs.
// Wire-format strings (snake_case) match what ScenarioConfigLoader expects.
// Edit only when the C# enums change.

export const MISSION_TYPES = [
  { value: 'hostage_rescue', label: 'Hostage Rescue' },
]

export const ROOM_SIZES = [
  { value: 'small',  label: 'Small (4x4 m)'  },
  { value: 'medium', label: 'Medium (6x6 m)' },
  { value: 'large',  label: 'Large (8x8 m)'  },
]

export const LAYOUT_TYPES = [
  { value: 'linear',        label: 'Linear'      },
  { value: 'branching',     label: 'Branching'   },
  { value: 'hub_and_spoke', label: 'Hub & Spoke' },
  { value: 'loop',          label: 'Loop'        },
]

export const ENTRY_TYPES = [
  { value: 'single',   label: 'Single'   },
  { value: 'multiple', label: 'Multiple' },
]

export const PLACEMENT_STRATEGIES = [
  { value: 'clustered',    label: 'Clustered'    },
  { value: 'dispersed',    label: 'Dispersed'    },
  { value: 'front_loaded', label: 'Front-Loaded' },
  { value: 'deep',         label: 'Deep'         },
]

export const HOSTAGE_RISK_LEVELS = [
  { value: 'low',    label: 'Low'    },
  { value: 'medium', label: 'Medium' },
  { value: 'high',   label: 'High'   },
]

export const RANDOMNESS_LEVELS = [
  { value: 'low',    label: 'Low'    },
  { value: 'medium', label: 'Medium' },
  { value: 'high',   label: 'High'   },
]

export const DIFFICULTY_LABELS = {
  1: '1 (Easiest)',
  2: '2 (Easy)',
  3: '3 (Moderate)',
  4: '4 (Hard)',
  5: '5 (Hardest)',
}

export const DEFAULT_CONFIG = {
  schemaVersion: '1.0.0',
  missionStructure: {
    missionType: 'hostage_rescue',
    roomCount:   { min: 4, max: 4 },
    roomSize:    'medium',
    layoutType:  'branching',
    entryType:   'single',
  },
  entityConfiguration: {
    hostageCount:      1,
    terroristCount:    4,
    placementStrategy: 'dispersed',
    hostageRiskLevel:  'medium',
  },
  executionControls: {
    difficultyLevel: 3,
    randomnessLevel: 'medium',
    seed:            null,
    timeLimit:       null,
    customLabel:     null,
  },
}
